### PR TITLE
feat(client-hmr): auto-reconnect on server disconnect

### PR DESCRIPTION
**Pull request by jinchaofs**
## Related Links

关于Rspress的一个issue: https://github.com/web-infra-dev/rspress/issues/464

问题原因是，当`_meta.json`文件更新时，会关闭`devServer`，然后再重启，这一步会导致跟hmr客户端断开连接，只能靠用户手动刷新。
image
image

解决方案是hmr客户端发现断开后，轮询尝试重连，最大轮询次数暂定100

### CODE REVIEW

**Comment by 9aoy** [comment]
thanks for your support, we will review it asap!

File: packages/core/src/client/hmr/index.ts [hunk 1]
```diff
@@ -216,3 +172,78 @@ function tryApplyUpdates() {
...
+
+  connection.onopen = function () {
+    isConnected = true;
+    if (typeof console !== 'undefined' && typeof console.info === 'function') {
+      // Notify users that the HMR has successfully connected.
+      console.info('[HMR] connected.');
+    }
+  };
+
+  // Unlike WebpackDevServer client, we won't try to reconnect
```
Comment (chenjiahan) [inline packages/core/src/client/hmr/index.ts:192]: This comment can be updated as Rsbuild will attempt to reconnect 😄

Comment (jinchaofs) [reply packages/core/src/client/hmr/index.ts]: OK😄

File: packages/core/src/client/hmr/index.ts [hunk 2]
```diff
@@ -216,3 +172,79 @@ function tryApplyUpdates() {
...
+
+    let retry_count = 0;
+    const retryInterval = setInterval(() => {
+      retry_count++;
+
+      if (isConnected || retry_count > MAX_RETRIES) {
+        clearInterval(retryInterval);
+        return;
+      }
+      reconnect();
```
Comment (chenjiahan) [inline packages/core/src/client/hmr/index.ts:208]: Is reconnect called too many times?

I tried to stop the local dev server, and the page creates many many requests:

image

Comment (jinchaofs) [reply packages/core/src/client/hmr/index.ts:208]: Oh, I'll take a look and see what's going on.

Comment (jinchaofs) [reply packages/core/src/client/hmr/index.ts:208]: When a reconnection attempt fails, it also triggers a 'close' event, leading to repeated setInterval calls. I'll address this issue in the fix~

File: packages/core/src/client/hmr/index.ts [hunk 3]
```diff
@@ -216,3 +172,110 @@ function tryApplyUpdates() {
     );
   }
 }
+
+const MAX_RETRIES = 20;
```
Comment (chenjiahan) [inline packages/core/src/client/hmr/index.ts:176]: Some large app may need more time to restart the dev server, so MAX_RETRIES = 100 can be better

```suggestion
const MAX_RETRIES = 100;
```

**Review by chenjiahan — APPROVED** [review]

**Comment by chenjiahan** [comment]
Thanks! This change greatly improves the development experience ❤️

**Pull request merged**